Packages/Tracking Preview/Locomotion/Runtime/Scripts/Teleportation/Teleport Action/Jump Gem Teleport/JumpGemTeleport.cs

```diff
@@ -0,0 +1,160 @@
+ * Copyright (C) Ultraleap, Inc. 2011-2022.                                   *
+ *                                                                            *
+ * Use subject to the terms of the Apache License 2.0 available at            *
+ * http://www.apache.org/licenses/LICENSE-2.0, or another agreement           *
+ * between Ultraleap and you, your company or other organization.             *
+ ******************************************************************************/
+
+using Leap.Unity.Preview.HandRays;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace Leap.Unity.Preview.Locomotion
+{
+    /// <summary>
+    /// This teleport method snaps a ray to a "gem" object within your scene when said gem is activated (e.g. through a pinch).
+    /// When the user releases the gem a teleport action is fired, if valid.
+    /// </summary>
+    public class JumpGemTeleport : TeleportActionBase
+    {
+        private List<JumpGem> _jumpGems = new List<JumpGem>();
+
+        public JumpGem PinchedGem => _pinchedGem;
+        private JumpGem _pinchedGem;
+
+        [Header("Jump Gem Teleport - Interaction Setup")]
+        [SerializeField, Range(0.0001f, 1f), Tooltip("The distance the user has to pinch a gem before the ray will be activated.")]
+        private float _minimumDistanceToActivate = 0.1f;
+
+        [SerializeField, Range(0f, 1f), Tooltip("The point at which the ray should interpolate when releasing. " +
+            "E.g. 0.8 will activate when the user is 80% released.")]
+        private float _releaseInterpolationPoint = 0.75f;
+
+        [SerializeField, Range(0f, 4f), Tooltip("The interpolation time that will be applied when releasing.")]
+        private float _releaseInterpolationTime = 2f;
+
+        /// <summary>
+        /// The distance the user has to pinch a gem before the ray will be activated.
+        /// </summary>
+        public float DistanceToActivate => _minimumDistanceToActivate;
+
+        /// <summary>
+        /// Called when the gem is pinched or unpinched - passes true if pinched, or false if released
+        /// </summary>
+        public Action<bool> OnGemPinched;
+
+        private void Awake()
+        {
+            _jumpGems = FindObjectsOfType<JumpGem>(true).ToList();
+            for (int i = 0; i < _jumpGems.Count; i++)
+            {
+                int j = i;
+                _jumpGems[i].OnPinch += (val) => { OnJewelPinch(val, j); };
+                _jumpGems[i].OnRelease += OnJewelRelease;
+            }
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+            ChangeVisibleJewels();
+            ProcessPinchedJewel();
+        }
+
+        /// <summary>
+        /// Hides all non-active gems.
+        /// </summary>
+        private void ChangeVisibleJewels()
+        {
+            bool anyPinched = _jumpGems.Any(x => x.IsPinched);
```

Comment:
Might be worth checking if Any allocates as it's being called per frame.

Reply:
I think this is fine as the list is only ever 2 in length (unless the dev has more than one set of Jump Gems in the scene)